cabal-install/Distribution/Client/CmdFreeze.hs

```diff
@@ -0,0 +1,164 @@
+{-# LANGUAGE NamedFieldPuns, RecordWildCards #-}
+
+-- | cabal-install CLI command: freeze
+--
+module Distribution.Client.CmdFreeze (
+    freezeAction,
+  ) where
+
+import Distribution.Client.ProjectPlanning
+         ( ElaboratedInstallPlan, rebuildInstallPlan )
+import Distribution.Client.ProjectConfig
+         ( ProjectConfig(..), ProjectConfigShared(..)
+         , commandLineFlagsToProjectConfig, writeProjectLocalFreezeConfig
+         , findProjectRoot )
+import Distribution.Client.ProjectPlanning.Types
+         ( ElaboratedConfiguredPackage(..) )
+import Distribution.Client.Targets
+         ( UserConstraint(..) )
+import Distribution.Solver.Types.ConstraintSource
+         ( ConstraintSource(..) )
+import Distribution.Client.DistDirLayout
+         ( defaultDistDirLayout, defaultCabalDirLayout )
+import Distribution.Client.Config
+         ( defaultCabalDir )
+import qualified Distribution.Client.InstallPlan as InstallPlan
+
+
+import Distribution.Package
+         ( PackageName, packageName, packageVersion )
+import Distribution.Version
+         ( VersionRange, thisVersion, unionVersionRanges )
+import Distribution.PackageDescription
+         ( FlagAssignment )
+import Distribution.Client.Setup
+         ( GlobalFlags, ConfigFlags(..), ConfigExFlags, InstallFlags )
+import Distribution.Simple.Setup
+         ( HaddockFlags, fromFlagOrDefault )
+import Distribution.Simple.Utils
+         ( die, notice )
+import Distribution.Verbosity
+         ( normal )
+
+import Data.Monoid
+import qualified Data.Map as Map
+import Data.Map (Map)
+import Control.Monad (unless)
+import System.FilePath
+
+
+-- | To a first approximation, the @freeze@ command runs the first phase of
+-- the @build@ command where we bring the install plan up to date, and then
+-- based on the install plan we write out a @cabal.project.freeze@ config file.
+--
+-- For more details on how this works, see the module
+-- "Distribution.Client.ProjectOrchestration"
+--
+freezeAction :: (ConfigFlags, ConfigExFlags, InstallFlags, HaddockFlags)
+             -> [String] -> GlobalFlags -> IO ()
+freezeAction (configFlags, configExFlags, installFlags, haddockFlags)
+             extraArgs globalFlags = do
+
+    unless (null extraArgs) $
+      die $ "'fetch' doesn't take any extra arguments: "
```

Comment:
Oops, I was also writing a fetch command at the same time. Got things mixed up apparently :-)